File: scanner_comp.cpp
#include "scanner_comp.h"
#include <string>

using namespace std;

Scanner::Scanner(const string& filename) : line(1), column(0), pos(0) {
    ifstream file(filename);
    if (file.is_open()) {
        string txt_content((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
        content.assign(txt_content.begin(), txt_content.end());
    } else {
        cerr << "Error al abrir el archivo" << endl;
    }
    file.close();
}

Token* Scanner::next_token() {
    if (is_eof()) {
        return nullptr;
    }

    int estado = 0;
    string term = "";

    while (true) {
        char current_char = next_char();
        column++;

        if (estado == 0) {
            if (is_char(current_char)) {
                term += current_char;
                estado = 1;
            } else if (is_digit(current_char)) {
                estado = 2;
                term += current_char;
            } else if (is_space(current_char)) {
                estado = 0;
            } else if (is_operator(current_char)) {
                term += current_char;
                Token* token = new Token();
                token->set_type(Token::Tk_operator);
                token->set_text(term);
                token->set_line(line);
                token->set_column(column - term.length());
                return token;
            } else {
                cout << "Simbolo no reconocido" << endl;
                break;
            }
        } else if (estado == 1) {
            if (is_char(current_char) || is_digit(current_char)) {
                estado = 1;
                term += current_char;
            } else if (is_space(current_char) || is_operator(current_char) || is_eof_char(current_char)) {
                if (!is_eof_char(current_char)) {
                    back();
                }
                Token* token = new Token();
                token->set_type(Token::Tk_identifier);
                token->set_text(term);
                token->set_line(line);
                token->set_column(column - term.length());
                return token;
            } else {
                cout << "Identificador mal formado" << endl;
                break;
            }
        } else if (estado == 2) {
            if (is_digit(current_char) || current_char == '.') {
                estado = 2;
                term += current_char;
            } else if (!is_char(current_char) || is_eof_char(current_char)) {
                if (!is_eof_char(current_char)) {
                    back();
                }
                Token* token = new Token();
                token->set_type(Token::Tk_number);
                token->set_text(term);
                token->set_line(line);
                token->set_column(column - term.length());
                return token;
            } else {
                cout << "Numero no reconocido" << endl;
                break;
            }
        }
    }
}

bool Scanner::is_digit(char c) {
    return '0' <= c && c <= '9';
}

bool Scanner::is_char(char c) {
    return ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z');
}

bool Scanner::is_operator(char c) {
    return c == '>' || c == '<' || c == '=' || c == '!' || c == '+' || c == '-' || c == '*' || c == '/';
}

bool Scanner::is_space(char c) {
    if (c == '\n' || c == '\r') {
        line++;
        column = 0;
    }
    return c == ' ' || c == '\t' || c == '\n' || c == '\r';
}

char Scanner::next_char() {
    if (is_eof()) {
        return '$';
    }
    pos++;
    return content[pos - 1];
}

bool Scanner::is_eof() {
    return pos >= content.size();
}

void Scanner::back() {
    pos--;
    column--;
}

bool Scanner::is_eof_char(char c) {
    return c == '\0';
}

File: scanner_comp.py

import codecs
from token_comp import Token

class Scanner:
    def __init__(self, filename):
        self.line = 1
        self.column = 0
        try:
            with codecs.open(filename, 'r', encoding='utf-8') as file:
                txt_content = file.read()
            self.content = list(txt_content)
            self.pos = 0
        except Exception as ex:
            print(ex)


    def next_token(self):
        if self.is_eof():
            return None
        
        estado = 0
        term = ""

        while True:
            current_char = self.next_char()
            self.column += 1

            if estado == 0:
                if self.is_char(current_char):
                    term += current_char
                    estado = 1
                elif self.is_digit(current_char):
                    estado = 2
                    term += current_char
                elif self.is_space(current_char):
                    estado = 0
                elif self.is_operator(current_char):
                    term += current_char
                    token = Token()
                    token.set_type(token.TK_OPERATOR)
                    token.set_text(term)
                    token.set_line(self.line)
                    token.set_column(self.column - len(term))
                    return token
                else: 
                    print("SIMBOLO NO RECONOCIDO")
                    break

            elif estado == 1:
                if self.is_char(current_char) or self.is_digit(current_char):
                    estado = 1
                    term += current_char # concatenación del termino. 
                elif self.is_space(current_char) or self.is_operator(current_char) or self.is_eof_char(current_char):
                    if not self.is_eof_char(current_char):
                        self.back()
                    token = Token()
                    token.set_type(Token.TK_IDENTIFIER)
                    token.set_text(term)
                    token.set_line(self.line)
                    token.set_column(self.column - len(term))
                    return token
                else:
                    print("IDENTIFICADOR MAL FORMADO")
                    break

            elif estado == 2:
                if self.is_digit(current_char) or current_char == '.':
                    estado = 2
                    term += current_char # concatenación que es igual a decir term = term + current_char
                elif not self.is_char(current_char) or self.is_eof_char(current_char):
                    if not self.is_eof_char(current_char):
                        self.back()
                    token = Token()
                    token.set_type(Token.TK_NUMBER)
                    token.set_text(term)
                    token.set_line(self.line)
                    token.set_column(self.column - len(term))
                    return token
                else:
                    print("NÚMERO NO RECONOCIDO")
                    break

    # a?c + 123 * 456 / 12.0 
    # "abc 6" -> 'a', 'b', 'c', ' ', '6'

    def is_digit(self, c):
        return '0' <= c <= '9'

    def is_char(self, c):
        return 'a' <= c <= 'z' or 'A' <= c <= 'Z'
    
    def is_operator(self, c):
        return c in ['>', '<', '=', '!', '+', '-', '*', '/']
    
    def is_space(self, c):
        if c in ['\n', '\r']:
            self.line += 1
            self.column = 0
        return c in [' ', '\t', '\n', '\r']
    
    def next_char(self):
        if self.is_eof():
            return '$'
        self.pos = self.pos + 1
        return self.content[self.pos - 1]
    
    def is_eof(self): # es el final del archivo?
        return self.pos >= len(self.content)
    
    def back(self):
        self.pos -= 1
        self.column -= 1

    def is_eof_char(self, c):
        return c == '\0'